flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemHAServices.java

```diff
@@ -0,0 +1,342 @@
+
+	private final Object lock = new Object();
+
+	/** Configuration */
+	private final Configuration configuration;
+
+	/** Job Registry */
+	private final RunningJobsRegistry runningJobsRegistry;
+
+	/** Store for arbitrary blobs */
+	private final BlobStoreService blobStoreService;
+
+	/** Shutdown flag */
+	private boolean shutdown;
+
+	/** Executor */
+	private final Executor executor;
+
+	/** Used for Mini cluster (local == true) */
+	private EmbeddedHaServices miniclusterURLResolver;
+
+	/** Used for cluster (local == false) */
+	private StandaloneHaServices clusterURLResolver;
+
+	private final boolean local;
+
+	/**
+	 * Creates a new services class for local usage.
+	 *
+	 * @param executor    		Current executor
+	 * @param configuration		Configuration
+	 * @param blobStoreService	Blob store service
+	 */
+	public FileSystemHAServices(
+			Executor executor,
+			Configuration configuration,
+			BlobStoreService blobStoreService) {
+
+		miniclusterURLResolver = new EmbeddedHaServices(executor);
+
+		this.executor = checkNotNull(executor);
+		this.blobStoreService = checkNotNull(blobStoreService, "blobStoreService");
+		this.configuration = configuration;
+		this.runningJobsRegistry = new StandaloneRunningJobsRegistry();
+
+		local = true;
+		shutdown = false;
+	}
+
+	/**
+	 * Creates a new services class for cluster usage.
+	 *
+	 * @param resourceManagerAddress    The fix address of the ResourceManager
+	 * @param dispatcherAddress    		The fix address of the Dispatcher
+	 * @param jobManagerAddress    		The fix address of the Job Manager
+	 * @param webMonitorAddress			The fix address of the Web Monitor
+	 * @param executor    				Current executor
+	 * @param configuration				Configuration
+	 * @param blobStoreService			Blob store service
+	 */
+	public FileSystemHAServices(
+			String resourceManagerAddress,
+			String dispatcherAddress,
+			String jobManagerAddress,
+			String webMonitorAddress,
+			Executor executor,
+			Configuration configuration,
+			BlobStoreService blobStoreService) {
+
+		clusterURLResolver = new StandaloneHaServices(resourceManagerAddress, dispatcherAddress, jobManagerAddress, webMonitorAddress);
+
+		this.executor = checkNotNull(executor);
+		this.blobStoreService = checkNotNull(blobStoreService, "blobStoreService");
+		this.configuration = configuration;
+		this.runningJobsRegistry = new StandaloneRunningJobsRegistry();
+
+		local = false;
+		shutdown = false;
+	}
+
+
+	// ------------------------------------------------------------------------
+	//  Services
+	// ------------------------------------------------------------------------
+
+	@Override
+	public LeaderRetrievalService getResourceManagerLeaderRetriever() {
+		if (local) {
+			return miniclusterURLResolver.getResourceManagerLeaderRetriever();
+		}
+		else {
+			return clusterURLResolver.getResourceManagerLeaderRetriever();
+		}
+	}
+
+	@Override
+	public LeaderRetrievalService getDispatcherLeaderRetriever() {
+		if(local) {
```

Comment:
IMHO polymorphism would make this code easier to read

Reply:
Fixing it now. Missed the common parent